java-client/deephaven/src/main/java/io/deephaven/client/impl/BarrageSubscriptionImpl.java

```diff
@@ -0,0 +1,306 @@
+import io.deephaven.client.impl.util.BarrageProtoUtil;
+import io.deephaven.client.impl.util.BarrageStreamReader;
+import io.deephaven.client.impl.util.BarrageUtil;
+import io.deephaven.db.tables.TableDefinition;
+import io.deephaven.db.util.liveness.ReferenceCountedLivenessNode;
+import io.deephaven.db.v2.sources.chunk.ChunkType;
+import io.deephaven.db.v2.utils.BarrageMessage;
+import io.deephaven.db.v2.utils.Index;
+import io.deephaven.grpc_api.util.ExportTicketHelper;
+import io.deephaven.internal.log.LoggerFactory;
+import io.deephaven.io.logger.Logger;
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.MethodDescriptor;
+import io.grpc.protobuf.ProtoUtils;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientCalls;
+import io.grpc.stub.ClientResponseObserver;
+import org.apache.arrow.flight.impl.Flight;
+import org.apache.arrow.flight.impl.FlightServiceGrpc;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.BitSet;
+
+public class BarrageSubscriptionImpl extends ReferenceCountedLivenessNode implements BarrageSubscription {
+    private static final Logger log = LoggerFactory.getLogger(BarrageSubscriptionImpl.class);
+
+    private final String logName;
+    private final int updateIntervalMs;
+    private final BarrageSubscriptionOptions options;
+    private final ClientCall<Flight.FlightData, BarrageMessage> call;
+
+    private Runnable performRelease;
+    private BarrageTable resultTable;
+
+    private boolean subscribed = false;
+    private volatile boolean connected = false;
+
+    /**
+     * Represents a BarrageSubscription.
+     *
+     * @param session the deephaven session that this export belongs to
+     * @param export the export to subscribe to
+     * @param options the transport level options for this subscription
+     * @param updateIntervalMs the requested update interval; typically unspecified to conform to server config
+     * @param performRelease a callback that is invoked when this subscription is closed/destroyed/garbage-collected
+     */
+    public BarrageSubscriptionImpl(
+            final DeephavenClientSession session, final Export export, final BarrageSubscriptionOptions options,
+            final int updateIntervalMs, @Nullable final Runnable performRelease) {
+        super(false);
+
+        this.logName = ExportTicketHelper.toReadableString(export.ticket(), "export.ticket()");
+        this.updateIntervalMs = updateIntervalMs;
+        this.options = options;
+        this.performRelease = performRelease;
+
+        // fetch the schema and convert to table definition
+        final Schema schema = session.getSchema(export);
+        final TableDefinition definition = BarrageUtil.schemaToTableDefinition(schema);
+
+        resultTable = BarrageTable.make(definition, false);
+        resultTable.addParentReference(this);
```

Comment:
I'd oh so like it if we could have the client optionally provide the TableDefinition (the qst equivalent is TableHeader, but it's not plumbed through the engine like TableDefinition is...) to avoid this call.

getSchema is also a blocking call, and I'm always suspicious of constructors that do real work... does it make sense for the constructor to take in `TableDefinition`; and the caller is responsible for getting the schema if necessary?